verify.js

```diff
@@ -86,96 +51,143 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
 
   var header = decodedToken.header;
 
-  if (!~options.algorithms.indexOf(header.alg)) {
-    return done(new JsonWebTokenError('invalid algorithm'));
+  if(header.alg === 'none' && decodedToken.signature) {
+    return done(new JsonWebTokenError('invalid token: unsigned but signature is present'));
   }
 
-  var valid;
+  var getSecret;
 
-  try {
-    valid = jws.verify(jwtString, header.alg, secretOrPublicKey);
-  } catch (e) {
-    return done(e);
+  if(typeof secretOrPublicKey !== 'function') {
+    getSecret = function(header, callback) {
+      return callback(null, secretOrPublicKey);
+    };
   }
+  else if(header.alg === 'none') {
+    getSecret = function(header, callback) {
+      return callback(null, undefined);
+    };
+  }
+  else {
+    if(!callback) {
+      return done(new JsonWebTokenError('verify must be called asynchronous if secret or public key is provided as a callback'));
+    }
 
-  if (!valid)
-    return done(new JsonWebTokenError('invalid signature'));
+    getSecret = secretOrPublicKey;
+  }
 
-  var payload;
+  return getSecret(header, function(err, secretOrPublicKey) {
```

Comment:
Again, don't shadow outer variable names.  Something that was passed as a function is suddenly a string within the same function...  (perhaps it is the upper variable that should be renamed in this case, but anyhow, don't reuse it for two different things)